ENH: improve duck typing inside iscomplexobj

goerz opened the pull request:
Both `iscomplexobj` and `isrealobj` now try to refer to the argument's
`dtype` attribute if it exists. This significantly extends the types for
which `iscomplexobj` returns correct results (to e.g. scipy sparse
matrices).

This fixes #7924

goerz commented:
As far as I can tell none of the failing tests are related to my modifications

shoyer commented on numpy/lib/type_check.py:
```diff
@@ -266,7 +272,19 @@ def iscomplexobj(x):
+            # The numpy.testing.utils.assert_equal routine relies on calling
+            # iscomplexobj with weird arguments (types) and catches ValueErrors
+            # only. We therefore convert AttributeError to ValueError to
+            # maintain a compatible interface.
```
Is this really part of the public API? Could we simply fix `assert_equal` instead?

This seems like an unfortunate work around to preserve for backwards compatibility.

shoyer commented on numpy/lib/type_check.py:
```diff
@@ -266,7 +272,19 @@ def iscomplexobj(x):
     """
-    return issubclass(asarray(x).dtype.type, _nx.complexfloating)
+    try:
+        dtype = x.dtype
```
If an exception does not get raised here, I'm pretty sure nothing gets returned from this function.

This is good example of why we insist on unit tests :)

goerz commented:
Yeah, I screwed up the indentation level :-( . I added some additional unit tests.

I completely agree the workaround is ugly, and if you think it's better to let the AttributeError be raised here and catch that in `assert_equal`, I'll be happy to do that.

shoyer commented:
Can you add a test that checks an actual duck-type? Preferably, something that can't be converted into a NumPy array?

A robust way to check dypes might be something like:
`isinstance(dtype, np.dtype) and issubclass(dtype.type, _nx.complexfloating)`

I don't think we can expect this check to pass on duck array types that don't reuse numpy dtypes. That would probably solve the `assert_equals` issue, too.

shoyer commented:
for example:
```python
class DummyComplexArray:
    @property
    def dtype(self):
        return np.dtype(complex)

dummy = DummyComplexArray()
```

shoyer commented:
I'm not quite sure the right solution is coerce if the input array has a `dtype` that isn't a NumPy dtype.

The one concrete example I can think of of non-numpy dtypes in wide use is pandas, which has its own dtypes for `Categorical` types (among others). They are not NumPy dtypes, but they are pretty close, and actually do define the `type` attribute. So maybe we want:
`issubclass(getattr(dtype, 'type', None), _nx.complexfloating)`

This would fully support duck dtypes then, as well as duck arrays.

shoyer commented:
This looks good to me, though I'll leave it open for a day or two for comments.

Could you please squash into a single commit?

goerz commented:
Yeah, this should work now. So should I squash locally and then force-push to my github fork? Will this do the right thing, without messing up this thread? I think github also gives the option to squash when applying the pull request, so I just want to double-check that it's ok for me to do locally (I never tried this before)

shoyer commented:
>  So should I squash locally and then force-push to my github fork? Will this do the right thing, without messing up this thread?

Yes, that's exactly right.

> I think github also gives the option to squash when applying the pull request, so I just want to double-check that it's ok for me to do locally (I never tried this before)

Yes, it does. I'm not sure if we're using that for NumPy yet, though (and we tend to be pretty particular about our commit messages).

goerz committed:
ENH: improve duck typing inside iscomplexobj

Both `iscomplexobj` and `isrealobj` now try to refer to the argument's
`dtype` attribute if it exists. This significantly extends the list of
types for which `iscomplexobj` returns correct results (including e.g.
scipy sparse matrices and pandas objects).

Extended the tests of the `iscomplexobj` routine for the following cases:
* simple scalars
* standard lists (test internal auto-conversion to numpy arrays)
* "Duck typing" for objects that define a dtype attribute (either
  referring to one of the existing numpy dtypes, or a custom dtype, as
  pandas does)

This fixes #7924

goerz commented:
Cool, looks like it worked

juliantaylor commented:
looks good to me

shoyer commented:
Thanks @goerz !

goerz commented:
Thanks for the code review, @shoyer! It's astonishing how even a seemingly trivial fix can take multiple iterations to get right!

shoyer commented:
I promise, it gets easier with practice :).

behrmann commented:
I might be a bit late to the game, but this new version of `iscomplexobj` has a small corner case: it only works if the dtype attribute of an object is Numpy-compatible, like the one Pandas has, and itself has a type attribute. This precludes using Python types, like `complex`, as a custom dtype. 

Is this intended or should I open an issue for that?

shoyer commented:
Yes, please open a new issue, preferably with a self-contained example of
the undesired behavior

On Fri, Feb 10, 2017 at 5:44 AM, Jörg Behrmann <notifications@github.com>
wrote:

> I might be a bit late to the game, but this new version of iscomplexobj
> has a small corner case: it only works if the dtype attribute of an object
> is Numpy-compatible, like the one Pandas has, and itself has a type
> attribute. This precludes using Python types, like complex, as a custom
> dtype.
>
> Is this intended or should I open an issue for that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/7936#issuecomment-278944935>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABKS1qRN1xckrbFRnUVYuwuJAnH6Bp6Bks5rbGm0gaJpZM4Jj_CK>
> .
>
